src/service_discovery/mod.rs

```diff
@@ -95,45 +104,38 @@ impl ServiceDiscovery {
     pub fn seek_peers(&mut self) -> Result<(), ServiceDiscoveryError> {
         let _ = self
             .socket
-            .send_to(&self.seek_peers_req, &self.remote_addr)?;
+            .write_to(Some((&self.seek_peers_req, self.remote_addr, 0)))?;
         Ok(())
     }
 
     /// Register service discovery observer
-    pub fn register_observer(&mut self, obs: Sender<Vec<SocketAddr>>) {
+    pub fn register_observer(&mut self, obs: Sender<Vec<PeerInfo>>) {
         self.observers.push(obs);
     }
 
     fn read(&mut self, core: &mut Core, poll: &Poll) {
-        let (bytes_rxd, peer_addr) = match self.socket.recv_from(&mut self.read_buf) {
-            Ok((bytes_rxd, peer_addr)) => (bytes_rxd, peer_addr),
-            // TODO(povilas): handle WouldBlock as well
-            Err(ref e) if e.kind() == ErrorKind::Interrupted => return,
-            Err(e) => {
-                debug!("ServiceDiscovery error in read: {:?}", e);
-                self.terminate(core, poll);
-                return;
-            }
-        };
-
-        let msg: DiscoveryMsg = match deserialise(&self.read_buf[..bytes_rxd]) {
-            Ok(msg) => msg,
-            Err(e) => {
-                debug!("Bogus message serialisation error: {:?}", e);
-                return;
-            }
-        };
-
-        match msg {
-            DiscoveryMsg::Request { guid } => {
-                if self.listen && self.guid != guid {
-                    self.reply_to.push_back(peer_addr);
-                    self.write(core, poll)
+        loop {
+            let (msg, peer_addr) = match self.socket.read_frm() {
+                Ok(Some((msg, peer_addr))) => (msg, peer_addr),
+                Ok(None) => return,
```

Comment:
Isn't this a wouldblock scenario and you should break if you have collected something in previous run of the loop (if collecting in a vector for e.g.) ?

Reply:
mhm and that's what it exactly does: breaks the loop and returns from the function.